### PR TITLE
update scalajs-dom to 2.0.0

### PR DESCRIPTION
**Pull request by hugo-vrijswijk**
Update scalajs-dom to 2.0.0. This version is not binary compatible with 1.2.0. I was running into an issue where some of my other dependencies had upgraded, but ST hadn't, resulting in an eviction error from sbt.

I've mostly followed the [migration guide](https://github.com/scala-js/scala-js-dom/releases/tag/v2.0.0):

1. Change version number
2. Remove  .for3Use2_13
3. Run script to migrate most imports
4. Run find-replace on `dom.raw`, `dom.experimental`, `dom.crypto` to replace with `dom` imports
5. Test on a local project

Two pre-release versions are used to be able to use scalajs-dom 2 on tests projects:

- [ ] japgolly react 2.0.0-RC4. Find-replace when final version is released
- [ ] slinky 0.6.8+1-b1dd3f4c. Find-replace when final version is released

I realize the changelog for this is _massive_. But because of the changed imports there's not really a way around it. When someone reviews this, perhaps there's a way to filter out certain diffs (like anything that starts with `scalajs.dom.`?

### CODE REVIEW

**Comment by rpiaggio** [comment]
> Snapshot published as `1.0.0-beta36+27-dc516c4d-SNAPSHOT`, it needs this resolver to work: `resolvers += MavenRepository("sonatype-s01-snapshots", "https://s01.oss.sonatype.org/content/repositories/snapshots")`

Would it be straightforward to release a snapshot depending on scalajs-react 2.0.0-RC5 ?

My deepest thanks to everyone involved.

**Comment by rpiaggio** [comment]
Hmm. I realize we would need a snapshot that also integrates this PR in order to be able to migrate to scalajs-dom 2.

**Comment by armanbilge** [comment]
@rpiaggio https://github.com/ScalablyTyped/Converter/pull/326 is based on this PR, so the snapshot published from there against scalajs-react 2.0.0-RC4 works with sjs-dom v2. In fact, I would have expected it to work for RC5 unless there are breaking changes?

**Comment by rpiaggio** [comment]
Thank you for the quick response!

> @rpiaggio #326 is based on this PR, so the snapshot published from there against scalajs-react 2.0.0-RC4 works with sjs-dom v2. In fact, I would have expected it to work for RC5 unless there are breaking changes?

I'm doing `ThisBuild / libraryDependencySchemes += "com.github.japgolly.scalajs-react" %% "core_sjs1" % "always"` to avoid the version conflict error betweekn RC4 and RC5; but then I get an error when compiling this ST generated class:

```
std/ParentNode.scala:305:17: class HTMLCollection takes type parameters
  val children: HTMLCollection = js.native
```

`HTMLCollection` is `org.scalajs.dom.HTMLCollection`.

We have `stUseScalaJsDom := true`, but setting it to `false` makes no difference.

**Comment by armanbilge** [comment]
Thanks for reporting, that's because of the changes in https://github.com/scala-js/scala-js-dom/pull/593. It looks like the generator in ST will have to be adjusted to accommodate for that.

I am surprised that setting `stUseScalaJsDom := false` doesn't work, though.

File: scalajs/src/main/scala/org/scalablytyped/converter/internal/scalajs/flavours/ScalaJsDomNames.scala [hunk 1]
```diff
@@ -6,543 +6,543 @@ import org.scalablytyped.converter.internal.scalajs.CastConversion.TParam._1
...
+      CastConversion(stdNames.lib + Name("IDBEnvironment"), QualifiedName("org.scalajs.dom.IDBEnvironment")),
+      CastConversion(stdNames.lib + Name("IDBFactory"), QualifiedName("org.scalajs.dom.IDBFactory")),
+      CastConversion(stdNames.lib + Name("IDBIndex"), QualifiedName("org.scalajs.dom.IDBIndex")),
+      CastConversion(stdNames.lib + Name("IDBKeyRange"), QualifiedName("org.scalajs.dom.IDBKeyRange")),
+      CastConversion(stdNames.lib + Name("IDBObjectStore"), QualifiedName("org.scalajs.dom.IDBObjectStore")),
+      CastConversion(stdNames.lib + Name("IDBOpenDBRequest"), QualifiedName("org.scalajs.dom.IDBOpenDBRequest")),
+      CastConversion(stdNames.lib + Name("IDBRequest"), QualifiedName("org.scalajs.dom.IDBRequest")),
+      CastConversion(stdNames.lib + Name("IDBTransaction"), QualifiedName("org.scalajs.dom.IDBTransaction")),
+      CastConversion(stdNames.lib + Name("IDBVersionChangeEvent"), QualifiedName("org.scalajs.dom.IDBVersionChangeEvent")),
       CastConversion(stdNames.lib + Name("Image"), QualifiedName("org.scalajs.dom.ext.Image")),
```
[the remark this answers is not in the document]
Comment (armanbilge) [reply scalajs/src/main/scala/org/scalablytyped/converter/internal/scalajs/flavours/ScalaJsDomNames.scala:219]: Made https://github.com/scala-js/scala-js-dom/issues/621 to track this, maybe we'll get it in the next release 😁

[the remark this answers is not in the document]
Comment (armanbilge) [reply scalajs/src/main/scala/org/scalablytyped/converter/internal/scalajs/flavours/ScalaJsDomNames.scala:219]:
```suggestion
```

**Comment by oyvindberg** [comment]
I wrote a test to enforce that all right hand sides in that conversion tables exist, and that the number of type parameters matches. Then I fixed the few things which showed up. The biggest surprise was probably that `IDBRequest` has a type parameter which TS doesn't.

For reference, these are the types in scala-js-dom which ST so far will not translate to. I don't intend to add any of those to the conversion list now, but anyone can certainly feel free

<details>
  <summary>somewhat long list</summary>

```
org.scalajs.dom.AppendMode
org.scalajs.dom.BigInteger
org.scalajs.dom.BufferSource
org.scalajs.dom.CanvasProxy
org.scalajs.dom.Chunk
org.scalajs.dom.Client
org.scalajs.dom.ClientRect
org.scalajs.dom.ClientRectList
org.scalajs.dom.ClientType
org.scalajs.dom.Clients
org.scalajs.dom.Clipboard
org.scalajs.dom.CompositionEventInit
org.scalajs.dom.ConvertToBlobOptions
org.scalajs.dom.CreateImageBitmapOptions
org.scalajs.dom.CustomEventInit
org.scalajs.dom.DOMList
org.scalajs.dom.DedicatedWorkerGlobalScope
org.scalajs.dom.DocumentReadyState
org.scalajs.dom.EndOfStreamError
org.scalajs.dom.EventException
org.scalajs.dom.EventInit
org.scalajs.dom.ExtendableEvent
org.scalajs.dom.ExtendableEventInit
org.scalajs.dom.ExtendableMessageEvent
org.scalajs.dom.ExtendableMessageEventInit
org.scalajs.dom.FetchEvent
org.scalajs.dom.FetchEventInit
org.scalajs.dom.FileReaderSync
org.scalajs.dom.FocusEventInit
org.scalajs.dom.FrameType
org.scalajs.dom.FullscreenOptions
org.scalajs.dom.GetSVGDocument
org.scalajs.dom.HTMLFormControlsCollection
org.scalajs.dom.HTMLOptionsCollection
org.scalajs.dom.HTMLTableAlignment
org.scalajs.dom.HashAlgorithm
org.scalajs.dom.HeadersInit
org.scalajs.dom.HttpMethod
org.scalajs.dom.IDBCreateIndexOptions
org.scalajs.dom.IDBCreateObjectStoreOptions
org.scalajs.dom.IDBCursorDirection
org.scalajs.dom.IDBCursorReadOnly
org.scalajs.dom.IDBDatabaseInfo
org.scalajs.dom.IDBEvent
org.scalajs.dom.IDBEventTarget
org.scalajs.dom.IDBRequestReadyState
org.scalajs.dom.IDBStoreLike
org.scalajs.dom.IDBTransactionDurability
org.scalajs.dom.IDBTransactionMode
org.scalajs.dom.IDBTransactionOptions
org.scalajs.dom.IDBVersionChangeEventInit
org.scalajs.dom.ImageBitmap
org.scalajs.dom.ImageCapture
org.scalajs.dom.MediaQueryListListener
org.scalajs.dom.MediaSource
org.scalajs.dom.MediaSourceReadyState
org.scalajs.dom.MediaStreamEventInit
org.scalajs.dom.MessageEventInit
org.scalajs.dom.ModifierKeyEvent
org.scalajs.dom.ModifierKeyEventInit
org.scalajs.dom.MouseEventInit
org.scalajs.dom.NavigatorGeolocation
org.scalajs.dom.NavigatorStorageUtils
org.scalajs.dom.NavigatorVibration
org.scalajs.dom.NodeListOf
org.scalajs.dom.NodeSelector
org.scalajs.dom.OffscreenCanvas
org.scalajs.dom.OpenEndedDictionary
org.scalajs.dom.PageTransitionEvent
org.scalajs.dom.PageVisibility
org.scalajs.dom.PermissionDescriptor
org.scalajs.dom.PermissionName
org.scalajs.dom.PermissionState
org.scalajs.dom.PermissionStatus
org.scalajs.dom.Permissions
org.scalajs.dom.PushEvent
org.scalajs.dom.PushMessageData
org.scalajs.dom.PushPermissionDescriptor
org.scalajs.dom.RadioNodeList
org.scalajs.dom.ReadableStreamController
org.scalajs.dom.RequestType
org.scalajs.dom.SVGAnimatedPathData
org.scalajs.dom.SVGException
org.scalajs.dom.SVGExternalResourcesRequired
org.scalajs.dom.SVGLangSpace
org.scalajs.dom.SVGLocatable
org.scalajs.dom.SVGStylable
org.scalajs.dom.SVGTransformable
org.scalajs.dom.SVGZoomEventInit
org.scalajs.dom.Sequence
org.scalajs.dom.ServiceWorkerGlobalScope
org.scalajs.dom.SharedWorker
org.scalajs.dom.SharedWorkerGlobalScope
org.scalajs.dom.SourceBuffer
org.scalajs.dom.SourceBufferList
org.scalajs.dom.StorageEstimate
org.scalajs.dom.StorageEventInit
org.scalajs.dom.StorageManager
org.scalajs.dom.TextEventInit
org.scalajs.dom.TextTrackMode
org.scalajs.dom.TouchEventInit
org.scalajs.dom.Transferable
org.scalajs.dom.TwoDContextAttributes
org.scalajs.dom.UIEventInit
org.scalajs.dom.URLSearchParams
org.scalajs.dom.VideoTrack
org.scalajs.dom.VideoTrackList
org.scalajs.dom.WheelEventInit
org.scalajs.dom.WindowClient
org.scalajs.dom.WindowOrWorkerGlobalScope
org.scalajs.dom.WorkerGlobalScope
org.scalajs.dom.WorkerLocation
org.scalajs.dom.WorkerNavigator
org.scalajs.dom.WriteableState
org.scalajs.dom.WriteableStream
org.scalajs.dom.XPathNSResolver
org.scalajs.dom.webgl.ActiveInfo
org.scalajs.dom.webgl.Buffer
org.scalajs.dom.webgl.ContextAttributes
org.scalajs.dom.webgl.Framebuffer
org.scalajs.dom.webgl.Program
org.scalajs.dom.webgl.Renderbuffer
org.scalajs.dom.webgl.RenderingContext
org.scalajs.dom.webgl.Shader
org.scalajs.dom.webgl.ShaderPrecisionFormat
org.scalajs.dom.webgl.Texture
org.scalajs.dom.webgl.UniformLocation
org.scalajs.dom.webgl.extensions.ANGLEInstancedArrays
org.scalajs.dom.webgl.extensions.EXTBlendMinmax
org.scalajs.dom.webgl.extensions.EXTColorBufferHalfFloat
org.scalajs.dom.webgl.extensions.EXTDisjointTimerQuery
org.scalajs.dom.webgl.extensions.EXTFragDepth
org.scalajs.dom.webgl.extensions.EXTShaderTextureLOD
org.scalajs.dom.webgl.extensions.EXTTextureFilterAnisotropic
org.scalajs.dom.webgl.extensions.EXTsRGB
org.scalajs.dom.webgl.extensions.OESElementIndexUint
org.scalajs.dom.webgl.extensions.OESStandardDerivatives
org.scalajs.dom.webgl.extensions.OESTextureFloat
org.scalajs.dom.webgl.extensions.OESTextureFloatLinear
org.scalajs.dom.webgl.extensions.OESTextureHalfFloat
org.scalajs.dom.webgl.extensions.OESTextureHalfFloatLinear
org.scalajs.dom.webgl.extensions.OESVertexArrayObject
org.scalajs.dom.webgl.extensions.WEBGLColorBufferFloat
org.scalajs.dom.webgl.extensions.WEBGLCompressedTextureETC1
org.scalajs.dom.webgl.extensions.WEBGLDrawBuffers
org.scalajs.dom.webgl.extensions.WebGLCompressedTextureATC
org.scalajs.dom.webgl.extensions.WebGLCompressedTexturePVRTC
org.scalajs.dom.webgl.extensions.WebGLCompressedTextureS3TC
org.scalajs.dom.webgl.extensions.WebGLDebugRendererInfo
org.scalajs.dom.webgl.extensions.WebGLDebugShaders
org.scalajs.dom.webgl.extensions.WebGLDepthTexture
org.scalajs.dom.webgl.extensions.WebGLLoseContext
org.scalajs.dom.webgl.extensions.WebGLTimerQueryEXT
org.scalajs.dom.webgl.extensions.WebGLVertexArrayObjectOES
org.scalajs.dom.webworkers.AbstractWorker
org.scalajs.dom.webworkers.DedicatedWorkerGlobalScope
org.scalajs.dom.webworkers.Worker
org.scalajs.dom.webworkers.WorkerGlobalScope
org.scalajs.dom.webworkers.WorkerLocation
org.scalajs.dom.webworkers.WorkerNavigator
```

</details>

I also bumped #326 to use 2.0.0-M5, a snapshot if hopefully on the way

**Comment by armanbilge** [comment]
> The biggest surprise was probably that IDBRequest has a type parameter which TS doesn't.

Yes, I think that was done in https://github.com/scala-js/scala-js-dom/pull/569. Hope it didn't cause too much trouble 😅 

Quite a list there!

**Comment by oyvindberg** [comment]
No, not a problem at all @armanbilge  :)

snapshot published with version `1.0.0-beta36+29-f3b7dd30-SNAPSHOT`

**Comment by hugo-vrijswijk** [comment]
Thanks for the extensive review and comments. I might have bitten off more than I can chew with this PR, as my SJS and ST experience are limited, and this PR grew **a lot**. I see the build is still red, but I'm not sure what the exact reason for it is and what still needs to be done for this PR to be merged 😅

**Comment by armanbilge** [comment]
@hugo-vrijswijk as previously mentioned this PR is entangled with #326 which IIUC means they need to be merged at the same time (this one cannot standalone). Given that #326 is rebased against this, the fact that it is green there is a good sign :)

**Comment by rpiaggio** [comment]
> snapshot published with version `1.0.0-beta36+29-f3b7dd30-SNAPSHOT`

Awesome, our library compiles with this version. Thank you very much!

**Comment by ngbinh** [comment]
scala.js react 2.0.0 final has been released with not much changes https://github.com/japgolly/scalajs-react/blob/master/doc/changelog/2.0.0-RCs.md#changes-in-between-rc5-and-200-final

**Comment by oyvindberg** [comment]
Now we only need a stable version of slinky :)

**Comment by oyvindberg** [comment]
Ok, tired of being blocked. This is going in anyway, and I'll release when possible. with a slinky snapshot version if necessary

**Pull request merged**